modules/tensor_mechanics/src/auxkernels/ProjectedStatefulMaterialAux.C

```diff
@@ -0,0 +1,68 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ProjectedStatefulMaterialAux.h"
+#include "NodalPatchRecoveryBase.h"
+#include "MaterialBase.h"
+
+registerMooseObject("TensorMechanicsApp", ProjectedStatefulMaterialAux);
+registerMooseObject("TensorMechanicsApp", ADProjectedStatefulMaterialAux);
+
+template <bool is_ad>
+InputParameters
+ProjectedStatefulMaterialAuxTempl<is_ad>::validParams()
+{
+  auto params = IndexableProperty<AuxKernel, is_ad>::validParams();
+  params.addClassDescription(
+      "Picks a component of an indexable material property to get projected on an elemental "
+      "Auxvariable. For use by ProjectedStatefulMaterialStorageAction.");
+  return params;
+}
+
+template <bool is_ad>
+ProjectedStatefulMaterialAuxTempl<is_ad>::ProjectedStatefulMaterialAuxTempl(
+    const InputParameters & parameters)
+  : AuxKernel(parameters), _prop(this), _current_subdomain_id(_assembly.currentSubdomainID())
+{
+}
+
+template <bool is_ad>
+void
+ProjectedStatefulMaterialAuxTempl<is_ad>::initialSetup()
+{
+  _prop.check();
+
+  // get all material classes that provide properties for this object
+  _all_materials = getSupplyerMaterials();
+}
+
+template <bool is_ad>
+void
+ProjectedStatefulMaterialAuxTempl<is_ad>::subdomainSetup()
+{
+  // get materials for the current subdomain
+  _active_materials.clear();
+  for (const auto & mat : _all_materials)
+    if (mat->hasBlocks(_current_subdomain_id))
+      _active_materials.push_back(mat);
+}
+
+template <bool is_ad>
+Real
+ProjectedStatefulMaterialAuxTempl<is_ad>::computeValue()
+{
+  if (_t_step == 0)
+    for (const auto & mat : _active_materials)
+      mat->initStatefulProperties(_qrule->size());
```

Comment:
I need to think more about this w.r.t. stateful restart/recover

Reply:
Ugh, excellent point. But would `_t_step` not be something >0 in that case?

Reply:
either way, I have amended the tests to hopefully work now, and made them go 4 steps (hopefully that enables proper recover testing).

Reply:
Happy to report that `Modules recover` passes... unfortunately that doe not include a step for tensor mechanics... what the...?!

Reply:
Recover works locally:
```
xxx@xxxxx:~/Programs/moose/modules/tensor_mechanics$ ./run_tests --recover --re projected
test:projected_stateful_materials.monomial_first_part1 ................................................. PART1
test:projected_stateful_materials.monomial_constant_part1 .............................................. PART1
test:projected_stateful_materials.lagrange_first_part1 ................................................. PART1
test:projected_stateful_materials.monomial_first ................................................ [recover] OK
test:projected_stateful_materials.monomial_constant ............................................. [recover] OK
test:projected_stateful_materials.lagrange_first ................................................ [recover] OK
--------------------------------------------------------------------------------------------------------------
Ran 6 tests in 13.2 seconds. Average test time 1.3 seconds, maximum test time 1.3 seconds.
6 passed, 0 skipped, 0 pending, 0 failed
```

Reply:
I'm talking about from a design standpoint. I've come across many issues with manual calls to `initStatefulProperties` (and even internal ones...) that cause issues with advanced restart of stateful materials

Reply:
I'll resolve this per [your comment below](https://github.com/idaholab/moose/pull/26054#pullrequestreview-1768941840).